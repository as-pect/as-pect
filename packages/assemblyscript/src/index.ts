export type ASC = typeof import("assemblyscript/dist/asc.js") ;
export type AS = typeof import("assemblyscript/dist/assemblyscript.js");
export type TRANSFORM = typeof import("assemblyscript/dist/transform.js");

// @ts-ignore: Adding an aspect global
if (!globalThis.__aspect) globalThis.__aspect = {};

// We need to garuntee we use only a single version of the compiler

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.asc) globalThis.__aspect.asc = await import("assemblyscript/dist/asc.js");
// @ts-ignore: Adding an aspect global for asc
export const asc: ASC = globalThis.__aspect.asc;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.assemblyscript) globalThis.__aspect.assemblyscript = await import("assemblyscript/dist/assemblyscript.js");
// @ts-ignore: Adding an aspect global for asc
export const assemblyscript: AS = globalThis.__aspect.assemblyscript;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.transform) globalThis.__aspect.transform = await import("assemblyscript/dist/transform.js");
// @ts-ignore: Adding an aspect global for asc
export const transform: TRANSFORM = globalThis.__aspect.transform;

