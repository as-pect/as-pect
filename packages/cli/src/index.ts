import { parse } from "configinator";
import { cliConfig, resolveOptionByName } from "./util/configuration";
import chalk from "chalk";
const pkg = require("../package.json");

export function asp(args: string[], readFileSync: (file: string, basename: string) => string | null): void {
  const configState = parse(args, cliConfig, {
    cwd: process.cwd(),
    readFileSync,
  });

  if (configState.diagnostics.length > 0) {
    for (const diag of configState.diagnostics) {
      console.log(chalk`{red [CLI Error]}: ${diag}`);
    }
    process.exit(1);
  }

  if (resolveOptionByName(configState, "version")) {
    console.log(chalk`{yellow [Version]}: ${pkg.version}`);
    process.exit(0);
  }

  if (!resolveOptionByName(configState, "nologo")) {
    const printAsciiArt = require("./util/asciiArt").printAsciiArt;
    printAsciiArt(pkg.version);
  }

  if (resolveOptionByName(configState, "types")) {
    const types = require("./types").types;
    types();
    process.exit(0);
  }

  if (resolveOptionByName(configState, "portable")) {
    const portable = require("./portable").portable;
    portable();
    process.exit(0);
  }

  process.exit(0);
}

if (typeof require != "undefined" && require.main == module) {
  const fs = require("fs");
  const path = require("path");

  asp(process.argv.slice(2), (file: string, basename: string): string | null => {
    const thePath = path.normalize(path.join(basename, file));
    try {
      return fs.readFileSync(thePath, "utf8");
    } catch (ex) {
      return null;
    }
  });
}