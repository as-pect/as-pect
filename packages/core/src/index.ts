export * from "./reporter/CombinationReporter";
export * from "./reporter/EmptyReporter";
export * from "./reporter/IReporter";
export * from "./reporter/SummaryReporter";
export * from "./reporter/VerboseReporter";
export * from "./test/IWarning";
export * from "./test/TestContext";
export * from "./test/TestNode";
export * from "./util/IAspectExports";
export * from "./util/IPerformanceConfiguration";
export * from "./util/ReflectedValue";
