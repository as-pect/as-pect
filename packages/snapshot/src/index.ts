export * from "./parser";
export * from "./test/Snapshot";
export * from "./test/SnapshotDiff";
export * from "./test/unparse";
