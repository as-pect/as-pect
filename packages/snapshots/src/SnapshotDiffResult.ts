import { Change } from "diff";

export const enum SnapshotDiffResultType {
  NoChange,
  Added,
  Removed,
  Different,
}

export class SnapshotDiffResult {
  type: SnapshotDiffResultType = SnapshotDiffResultType.NoChange;
  left: string | null = null;
  right: string | null = null;
  changes: Change[] = [];
}
