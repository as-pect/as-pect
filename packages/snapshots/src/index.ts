export * from "./Snapshot";
export * from "./SnapshotDiff";
export * from "./SnapshotDiffResult";
