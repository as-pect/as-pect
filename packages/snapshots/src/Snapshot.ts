import { SnapshotDiff } from "./SnapshotDiff";
import { Parser, Grammar } from "nearley";
import grammar from "./parser/grammar";

const tick = /`/g;

export class Snapshot {
  public static parse(input: string): Snapshot {
    const parser = new Parser(Grammar.fromCompiled(grammar));
    parser.feed(input.replace(/\r/g, ""));
    if (parser.results.length !== 1)
      throw new Error("Ambiguous grammar or parsing.");
    const result = new Snapshot();
    result.values = parser.results[0];
    return result;
  }

  public static from(input: Map<string, string>): Snapshot {
    const snapshot = new Snapshot();
    snapshot.values = input;
    return snapshot;
  }

  values = new Map<string, string>();

  public add(key: string, value: string): this {
    let i = 0;
    while (true) {
      const snapshotKey = `${key}[${i}]`;
      if (!this.values.has(snapshotKey)) {
        this.values.set(snapshotKey, value);
        return this;
      }
      i++;
    }
  }

  public diff(other: Snapshot): SnapshotDiff {
    return new SnapshotDiff(this, other);
  }

  public stringify(): string {
    return (
      Array.from(this.values.entries())
        .map(
          ([key, value]) =>
            `exports[\`${key.replace(tick, "\\`")}\`] = \`${value.replace(
              tick,
              "\\`",
            )}\`;`,
        )
        .join("\n\n") + "\n"
    );
  }
}
