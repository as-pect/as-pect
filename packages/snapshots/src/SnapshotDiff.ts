import { Snapshot } from "./Snapshot";
import {
  SnapshotDiffResult,
  SnapshotDiffResultType,
} from "./SnapshotDiffResult";
import { diffLines } from "diff";

export class SnapshotDiff {
  results = new Map<string, SnapshotDiffResult>();

  constructor(public left: Snapshot, public right: Snapshot) {
    this.calculateDiff();
  }

  public calculateDiff(): void {
    const left = this.left.values;
    const right = this.right.values;

    // loop over the items on the left side
    for (const [key, value] of left.entries()) {
      // the snapshot exists, NoChange or Different
      if (right.has(key)) {
        const rightValue = right.get(key)!;
        const lines = diffLines(rightValue, value);
        const result = new SnapshotDiffResult();
        result.left = value;
        result.right = rightValue;
        result.type =
          value === rightValue
            ? SnapshotDiffResultType.NoChange
            : SnapshotDiffResultType.Different;
        result.changes = lines;
        this.results.set(key, result);
      } else {
        // it was added
        const result = new SnapshotDiffResult();
        result.left = value;
        result.right = null;
        result.type = SnapshotDiffResultType.Added;
        result.changes = diffLines("", value);
        this.results.set(key, result);
      }
    }

    // loop over the items on the right side
    for (const [key, value] of right.entries()) {
      if (!left.has(key)) {
        // it was removed
        const result = new SnapshotDiffResult();
        result.left = null;
        result.right = value;
        result.changes = diffLines(value, "");
        result.type = SnapshotDiffResultType.Removed;
        this.results.set(key, result);
      }
    }
  }
}
