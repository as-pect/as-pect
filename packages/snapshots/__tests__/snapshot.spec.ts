import { Snapshot } from "../src";

const inputA = `
exports[\`A\`] = \`SomeA {
      a: 1,
      b: 2,
      c: 3
    }\`;

exports[\`B\`] = \`SomeB {
      d: 4,
      e: 5,
      f: 6
    }\`;

exports[\`C\`] = \`SomeC {
      g: 7,
      h: 8,
      i: 9
    }\`;
`;
const inputB = `
exports[\`A\`] = \`SomeA {
      a: 1,
      b: 2,
      c: 3
    }\`;

exports[\`C\`] = \`SomeC {
      g: 4,
      h: 5,
      i: 6
    }\`;

exports[\`D\`] = \`SomeB {
      d: 1,
      e: 2,
      f: 3
    }\`;
`;

const map = new Map<string, string>();

describe("Snapshot", () => {
  it("should be instanceof Snapshot", () => {
    expect(new Snapshot()).toBeInstanceOf(Snapshot);
  });

  it("should parse a snapshot file", () => {
    expect(Snapshot.parse(inputA)).toMatchSnapshot();
  });

  it("should stringify a given snapshot", () => {
    expect(Snapshot.from(map).stringify()).toMatchSnapshot();
  });

  it("should diff snapshots", () => {
    expect(
      Snapshot.parse(inputA).diff(Snapshot.parse(inputB)),
    ).toMatchSnapshot();
  });
});
