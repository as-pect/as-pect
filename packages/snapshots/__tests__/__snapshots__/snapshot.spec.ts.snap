// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should diff snapshots 1`] = `
SnapshotDiff {
  "left": Snapshot {
    "values": Map {
      "A" => "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
      "B" => "SomeB {
      d: 4,
      e: 5,
      f: 6
    }",
      "C" => "SomeC {
      g: 7,
      h: 8,
      i: 9
    }",
    },
  },
  "results": Map {
    "A" => SnapshotDiffResult {
      "changes": Array [
        Object {
          "count": 5,
          "value": "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
        },
      ],
      "left": "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
      "right": "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
      "type": 0,
    },
    "B" => SnapshotDiffResult {
      "changes": Array [
        Object {
          "added": true,
          "count": 5,
          "removed": undefined,
          "value": "SomeB {
      d: 4,
      e: 5,
      f: 6
    }",
        },
      ],
      "left": "SomeB {
      d: 4,
      e: 5,
      f: 6
    }",
      "right": null,
      "type": 1,
    },
    "C" => SnapshotDiffResult {
      "changes": Array [
        Object {
          "count": 1,
          "value": "SomeC {
",
        },
        Object {
          "added": undefined,
          "count": 3,
          "removed": true,
          "value": "      g: 4,
      h: 5,
      i: 6
",
        },
        Object {
          "added": true,
          "count": 3,
          "removed": undefined,
          "value": "      g: 7,
      h: 8,
      i: 9
",
        },
        Object {
          "count": 1,
          "value": "    }",
        },
      ],
      "left": "SomeC {
      g: 7,
      h: 8,
      i: 9
    }",
      "right": "SomeC {
      g: 4,
      h: 5,
      i: 6
    }",
      "type": 3,
    },
    "D" => SnapshotDiffResult {
      "changes": Array [
        Object {
          "added": undefined,
          "count": 5,
          "removed": true,
          "value": "SomeB {
      d: 1,
      e: 2,
      f: 3
    }",
        },
      ],
      "left": null,
      "right": "SomeB {
      d: 1,
      e: 2,
      f: 3
    }",
      "type": 2,
    },
  },
  "right": Snapshot {
    "values": Map {
      "A" => "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
      "C" => "SomeC {
      g: 4,
      h: 5,
      i: 6
    }",
      "D" => "SomeB {
      d: 1,
      e: 2,
      f: 3
    }",
    },
  },
}
`;

exports[`Snapshot should parse a snapshot file 1`] = `
Snapshot {
  "values": Map {
    "A" => "SomeA {
      a: 1,
      b: 2,
      c: 3
    }",
    "B" => "SomeB {
      d: 4,
      e: 5,
      f: 6
    }",
    "C" => "SomeC {
      g: 7,
      h: 8,
      i: 9
    }",
  },
}
`;

exports[`Snapshot should stringify a given snapshot 1`] = `
"
"
`;
